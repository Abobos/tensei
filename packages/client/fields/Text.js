import React from 'react'
import { TextField } from '@contentful/forma-36-react-components'

class Text extends React.Component {
    render() {
        const { value, onFieldChange, field, errorMessage } = this.props

        return (
            <TextField
                value={(value || '').toString()}
                id={field.inputName}
                name={field.inputName}
                name={field.inputName}
                labelText={field.name}
                textInputProps={{
                    type: [
                        'NumberField',
                        'IntegerField',
                        'BigIntegerField'
                    ].includes(field.component)
                        ? 'number'
                        : 'text',
                    ...field.attributes
                }}
                validationMessage={errorMessage}
                description={field.description}
                onChange={event => onFieldChange(event.target.value)}
            />
        )
    }
}

export default Text
