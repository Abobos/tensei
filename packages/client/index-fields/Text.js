import React from 'react'

class Text extends React.Component {
    render() {
        const { value, ...rest } = this.props

        return value || ''
    }
}

export default Text
