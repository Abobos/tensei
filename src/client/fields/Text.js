import React from 'react'
import { TextField } from 'office-ui-fabric-react/lib/TextField'

class Text extends React.Component {
    render() {
        const { value, onChange, field, ...rest } = this.props

        return (
            <TextField
                {...rest}
                value={value}
                onChange={onChange}
                name={field.inputName}
                description={field.description}
            />
        )
    }
}

export default Text
