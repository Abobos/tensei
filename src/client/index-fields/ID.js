import React from 'react'
import TextTruncate from 'react-text-truncate'

class Text extends React.Component {
    render() {
        const { value, ...rest } = this.props

        return null
        // return value
    }
}

export default Text
