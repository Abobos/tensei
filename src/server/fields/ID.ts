import Field from './Field'

interface Constructor<M> {
    new (...args: any[]): M
}

class ID extends Field {
    /**
     *
     * If set to true, this field will
     * be cast to a string before
     * performing database
     * operations
     */
    public string: boolean = false

    /**
     *
     * With this set to true, this field will be
     * cast to a mongo object id before
     * performing database operations
     */
    public objectId: boolean = true

    /**
     * When a new ID field is created, by default,
     * we'll call the exceptOnForms() method.
     * That way, this field won't be
     * available on create
     * and update forms.
     */
    public constructor(name: string, databaseField?: string) {
        super(name, databaseField || '_id')

        this.exceptOnForms()
    }

    /**
     *
     * Set Id type to be string
     */
    public asString() {
        this.string = true
        this.objectId = false

        return this
    }

    /**
     *
     * Set iD type to be object id
     */
    public asObjectId() {
        this.string = false
        this.objectId = true

        return this
    }

    /**
     * Create a new instance of the field
     * requires constructor parameters
     *
     */
    public static make<T extends Field>(this: Constructor<T>, name?: string, databaseField?: string): T {
        return new this(name || 'ID', databaseField || '_id')
    }

    /**
     *
     * Add custom fields to the
     * serialize method
     */
    public serialize() {
        return {
            ...super.serialize(),
            asString: this.string,
            asObjectId: this.objectId,
        }
    }
}

export default ID
