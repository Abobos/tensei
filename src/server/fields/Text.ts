import Field from './Field'

class Text extends Field {
    /**
     *
     * Set a prefix to the text input on
     * the backend
     *
     */
    public withPrefix: string = ''

    /**
     *
     * Set a suffix to the text input on
     * the frontend
     */
    public withSuffix: string = ''

    /**
     *
     * An icon to append at the end of the
     * input
     *
     */
    public withIcon: string = ''

    /**
     * Set a prefix to the text input on
     * the backend
     *
     */
    public prefix(prefix: string) {
        this.withPrefix = prefix

        return this
    }

    /**
     *
     * Set a suffix to the text input on
     * the frontend
     */
    public suffix(suffix: string) {
        this.withSuffix = suffix

        return
    }

    /**
     *
     * Set an icon to be appended to input on
     * input on the frontend
     *
     */
    public icon(icon: string) {
        this.withIcon = icon

        return
    }

    /**
     *
     * Customize the serialize function to add
     * the suffix and prefix parameters
     */
    public serialize() {
        return {
            ...super.serialize(),

            prefix: this.withPrefix,
            suffix: this.withSuffix,
        }
    }
}

export default Text
