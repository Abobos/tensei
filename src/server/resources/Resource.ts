// import { Resource as ResourceInterface } from '../typings/interfaces'
import Pluralize from 'pluralize'
import { paramCase, capitalCase } from 'change-case'

class Resource {
    /**
     * This is the collection this resource will connect to
     * By default, it is the plural of the lower
     * case of the collection name.
     */
    public collection(): string {
        return Pluralize(paramCase(this.name()))
    }

    /**
     *
     * This is the primary key to be used to identify the resource
     * When resources are selected on the panel for example,
     * the primary key is used to identify them.
     */
    public primaryKey(): string {
        return '_id'
    }

    /**
     *
     * Use this to group resources on the left sidebar.
     * By default all resources are
     * in the `All` group
     */
    public group(): string {
        return 'All'
    }

    /**
     *
     * Define all the options for items to be fetched per page.
     * This would impact the number of items fetched on
     * page load.
     */
    public perPageOptions(): Array<number> {
        return [10, 25, 50, 100]
    }

    /**
     *
     * Determine if a resource should show up
     * on the left navigation.
     */
    public displayInNavigation(): boolean {
        return true
    }

    /**
     *
     * Get the name of the resource.
     */
    public name(): string {
        return this.constructor.name
    }

    /**
     *
     * This is the displayable label of the resource
     * It will appear on the left navigation
     * bar
     */
    public label(): string {
        return capitalCase(Pluralize(this.name()))
    }

    /**
     *
     * This will be used as the route param for
     * /resources/:param or /resuces/param
     */
    private param(): string {
        return Pluralize(paramCase(this.name()))
    }

    /**
     * Serialize the resource to be sent to
     * the frontend
     *
     */
    public serialize(): any {
        return {
            name: this.name(),
            label: this.label(),
            group: this.group(),
            param: this.param(),
            primaryKey: this.primaryKey(),
            collection: this.collection(),
            perPageOptions: this.perPageOptions(),
            displayInNavigation: this.displayInNavigation(),
        }
    }
}

export default Resource
