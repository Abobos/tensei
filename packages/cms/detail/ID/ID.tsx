import React, { EventHandler, SyntheticEvent } from 'react'
import { DetailComponentProps } from '@tensei/components'

import Text from '../Text'

const ID: React.FC<
  DetailComponentProps & {
    className?: string
    onClick?: EventHandler<SyntheticEvent<HTMLParagraphElement>>
  }
> = ({ value, values, field, resource, className = '' }) => {
  return (
    <>
      <Text
        field={field}
        value={value}
        values={values}
        resource={resource}
        className={`${className} bg-opacity-10 bg-tensei-primary text-xs rounded-md px-2 py-1 text-tensei-primary border border-tensei-primary w-auto inline-block`}
      ></Text>
    </>
  )
}

export default ID
