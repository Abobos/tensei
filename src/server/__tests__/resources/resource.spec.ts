import ID from '../../fields/ID'
import Text from '../../fields/Text'
import Resource from '../../resources/Resource'

class Post extends Resource {}

class User extends Resource {
    displayInNavigation() {
        return false
    }

    group() {
        return 'Finance'
    }

    label() {
        return 'Department leads'
    }

    primaryKey() {
        return 'id'
    }

    public perPageOptions() {
        return [100, 250, 500]
    }
}

class PostAuthorLabel extends Resource {}

class ShoppingCart extends Resource {
    public fields() {
        return [
            ID.make().asObjectId(),
            Text.make('Name', 'first_name')
                .sortable()
                .prefix('ox')
                .suffix('ox')
                .default('beans')
                .htmlAttributes({
                    required: true,
                    email: true,
                    title: 'User first name'
                })
                .hideFromIndex()
                .hideWhenUpdating()
        ]
    }
}

describe('Resource class', () => {
    it('Correctly serializes the resource', () => {
        const post = new Post()

        expect(post.serialize()).toEqual({
            label: 'Posts',
            collection: 'posts',
            displayInNavigation: true,
            group: 'All',
            primaryKey: '_id',
            param: 'posts',
            perPageOptions: [10, 25, 50, 100],
            name: 'Post',
            fields: []
        })
    })

    it('Correctly serializes the resource when properties are updated', () => {
        const user = new User()

        expect(user.serialize()).toEqual({
            label: 'Department leads',
            collection: 'users',
            displayInNavigation: false,
            group: 'Finance',
            primaryKey: 'id',
            perPageOptions: [100, 250, 500],
            name: 'User',
            param: 'users',
            fields: []
        })
    })

    it('serialises multiple word resources correctly', () => {
        const postAuthorLabel = new PostAuthorLabel()

        expect(postAuthorLabel.serialize()).toEqual({
            label: 'Post Author Labels',
            collection: 'post-author-labels',
            displayInNavigation: true,
            group: 'All',
            primaryKey: '_id',
            param: 'post-author-labels',
            perPageOptions: [10, 25, 50, 100],
            name: 'PostAuthorLabel',
            fields: []
        })

        expect(postAuthorLabel.serialize()).toMatchSnapshot()
    })

    it('correctly serializes all fields passed to resource', () => {
        const shoppingCart = new ShoppingCart()

        expect(shoppingCart.serialize()).toEqual({
            label: 'Shopping Carts',
            collection: 'shopping-carts',
            displayInNavigation: true,
            group: 'All',
            primaryKey: '_id',
            param: 'shopping-carts',
            perPageOptions: [10, 25, 50, 100],
            name: 'ShoppingCart',
            fields: [{
                attributes: {},
                component: 'id-field',
                databaseField: '_id',
                defaultValue: '',
                name: 'ID',
                showOnCreation: false,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: false,
                type: 'IDField',
                isSortable: false,
                asObjectId: true,
                asString: false
            }, {
                attributes: {
                    required: true,
                    email: true,
                    title: 'User first name'
                },
                component: 'text-field',
                databaseField: 'first_name',
                defaultValue: 'beans',
                name: 'Name',
                prefix: 'ox',
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: false,
                showOnUpdate: false,
                suffix: 'ox',
                type: 'TextField',
                isSortable: true
            }]
        })

        expect(shoppingCart.serialize()).toMatchSnapshot()
    })
})
