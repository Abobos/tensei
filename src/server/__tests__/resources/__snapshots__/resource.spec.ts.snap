// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource class correctly serializes all fields passed to resource 1`] = `
Object {
  "collection": "shopping-carts",
  "displayInNavigation": true,
  "fields": Array [
    Object {
      "asObjectId": true,
      "asString": false,
      "attributes": Object {},
      "component": "id-field",
      "databaseField": "_id",
      "defaultValue": "",
      "isSortable": false,
      "name": "ID",
      "showOnCreation": false,
      "showOnDetail": true,
      "showOnIndex": true,
      "showOnUpdate": false,
      "type": "IDField",
    },
    Object {
      "attributes": Object {
        "email": true,
        "required": true,
        "title": "User first name",
      },
      "component": "text-field",
      "databaseField": "first_name",
      "defaultValue": "beans",
      "isSortable": true,
      "name": "Name",
      "prefix": "ox",
      "showOnCreation": true,
      "showOnDetail": true,
      "showOnIndex": false,
      "showOnUpdate": false,
      "suffix": "ox",
      "type": "TextField",
    },
  ],
  "group": "All",
  "label": "Shopping Carts",
  "name": "ShoppingCart",
  "param": "shopping-carts",
  "perPageOptions": Array [
    10,
    25,
    50,
    100,
  ],
  "primaryKey": "_id",
}
`;

exports[`Resource class serialises multiple word resources correctly 1`] = `
Object {
  "collection": "post-author-labels",
  "displayInNavigation": true,
  "fields": Array [],
  "group": "All",
  "label": "Post Author Labels",
  "name": "PostAuthorLabel",
  "param": "post-author-labels",
  "perPageOptions": Array [
    10,
    25,
    50,
    100,
  ],
  "primaryKey": "_id",
}
`;
